host/class/hid/usb_host_hid/hid_host.c

```diff
@@ -424,50 +461,42 @@ static void hid_host_notify_interface_connected(hid_device_t *hid_device)
 static esp_err_t hid_host_interface_list_create(hid_device_t *hid_device,
         const usb_config_desc_t *config_desc)
 {
+    assert(hid_device);
+    assert(config_desc);
     size_t total_length = config_desc->wTotalLength;
     const usb_intf_desc_t *iface_desc = NULL;
     const hid_descriptor_t *hid_desc = NULL;
-    const usb_ep_desc_t *ep_desc = NULL;
     const usb_ep_desc_t *ep_in_desc = NULL;
-    int offset = 0;
+    int iface_offset = 0;
+    int hid_desc_offset = 0;
 
+    // Get first Interface descriptor
+    iface_desc = GET_NEXT_INTERFACE_DESC(config_desc, total_length, iface_offset);
     // For every Interface
-    for (int i = 0; i < config_desc->bNumInterfaces; i++) {
-        iface_desc = usb_parse_interface_descriptor(config_desc, i, 0, &offset);
+    while (iface_desc != NULL) {
+
         hid_desc = NULL;
+        hid_desc_offset = iface_offset;
         ep_in_desc = NULL;
 
         if (USB_CLASS_HID == iface_desc->bInterfaceClass) {
-            // HID descriptor
-            hid_desc = (const hid_descriptor_t *) usb_parse_next_descriptor_of_type((const usb_standard_desc_t *) iface_desc,
-                       total_length,
-                       HID_CLASS_DESCRIPTOR_TYPE_HID,
-                       &offset);
-            if (!hid_desc) {
-                return ESP_ERR_NOT_FOUND;
-            }
-
-            // EP descriptors for Interface
-            for (int i = 0; i < iface_desc->bNumEndpoints; i++) {
-                int ep_offset = 0;
-                ep_desc = usb_parse_endpoint_descriptor_by_index(iface_desc, i, total_length, &ep_offset);
-
-                if (ep_desc) {
-                    if (USB_EP_DESC_GET_EP_DIR(ep_desc)) {
-                        ep_in_desc = ep_desc;
-                    }
-                } else {
-                    return ESP_ERR_NOT_FOUND;
+            ESP_LOGD(TAG, "Found HID, bInterfaceNumber=%d, offset=%d, total_len=%d",
+                     iface_desc->bInterfaceNumber,
+                     iface_offset,
+                     total_length);
+            hid_desc = GET_NEXT_HID_DESC(iface_desc, total_length, hid_desc_offset);
```

Comment:
How are we sure that the 'next HID' descriptor belongs to this interface?
Could it happen that we have CFG descriptor like this:
```
intf0 desc
intf1 desc
   -hid0 desc
```
so we call `GET_NEXT_HID_DESC` on intf0, we actually get hid0, that belongs to hid1.

If that is not possible, feel free to disregard this comment

Reply:
@tore-espressif In your example intf0 - is a HID Class interface?

Reply:
> @tore-espressif In your example intf0 - is a HID Class interface?

inft0 means Interface descriptor with class set to HID.
hid0 means HID class specific descriptor

> UPD: I'am not sure that this is "legally" - to have HID interface without HID descriptor

Yes, that is basically the question. We must make sure that `GET_NEXT_HID_DESC(iface_desc,` returns 'HID_DESC' that belong to input argument 'iface_desc'.

Reply:
And is it legal for 1 Intf desc to have 2 (or more) HID descriptors?

Reply:
Anyway, good catch. It will fail in other part. 
I'll fix that.

Reply:
Here is something similar in CDC driver: https://github.com/espressif/idf-extra-components/blob/66994c5010096417f86a4328ae161424c45bd1ea/usb/usb_host_cdc_acm/cdc_acm_host.c#L727

I added it because I hit the problem with legal CDC devices

Reply:
> And is it legal for 1 Intf desc to have 2 (or more) HID descriptors?

No, there can be optional descriptor, which extends the HID descriptor.

Reply:
@tore-espressif, The position of HID descriptor is determined by specification, it shall be present and takes place in between Interface and Endpoint descriptors. 

![image](https://github.com/espressif/esp-usb/assets/11456035/d4c0d2f4-d20e-4768-83b8-ec5d4ed646da)